### ISSUE TITLE
Remove MissionType

### ISSUE THREAD

**Issue by bevans2000** [mars-sim/mars-sim#877]
Remove the MissionType so that new Missions can be added easily. The MissionType is a fixed value and  utilised in places through the code.
In particular it is important for defining how the UI works. This logic should be reworked to be based on a small set of Interfaces that capture the behaviour of Missions. There is already one for VehicleMission so this could be extended to ResourceMission, ConstructionMission & CommerceMission.

**Comment by bevans2000** [comment]
This will allow new Missions to be added more easily. The goal is to support different Missions of the same type but different implementations; this would allow a parallel running of the old & new mission implementation

**Comment by mokun** [comment]
So does it mean that we are working toward at some point in future, player may create a custom mission of his own ?

**Comment by bevans2000** [comment]
Yes  that is my plan. Once we have the MissionStep phase implemented we can create mission. that do multiple activities instead if being fixed. I need to work out how to select the custom UI screens

**Comment by mokun** [comment]
> that do multiple activities instead if being fixed.

Does it mean inside group activities that involve more than one settlers ?

**Comment by bevans2000** [comment]
Potentially, short term is to remove the type constraints so we have more freedom. We might need to go further and model a MissionOutcome concept that could hold the resources mined or collected as well as the cargo for commerce missions. This would remove the fixed nature in the current solution that drives the custom ui screens.
Essentially the Mission becomes generic with a step of custom steps and a set of Outcomes. For each Outcome type there is a dedicated screen; we have the screens now but they are tied to the Mission subclass.
Anyway I'm doing it on a branch so as not to break anything

**Comment by mokun** [comment]
> There is already one for VehicleMission so this could be extended to ResourceMission, ConstructionMission & CommerceMission

Construction mission is unique among other missions because it doesn't go anywhere. Everything happens in the vicinity of one settlement.

It would be interesting one day when we create a road network that will connect two nearby settlement together.

Commerce mission is also different from others in that it interacts with other settlements, thus requiring a thread to exchange data with another thread, which as you know can be very messy when it comes to reading and writing mutable data.

Yes making UI very flexible and configurable is the key. 

Say when we want to add a new type of mission, it shouldn't be too hard for us developers. 

Hopefully it only takes a few steps to assemble a few UI elements together for this new mission.

**Comment by mokun** [comment]
> This would remove the fixed nature in the current solution that drives the custom ui screens.

Yes making UI elements composable, flexible and configurable seems to be one key to make new mission easy to create.

Say when we want to add a new type of mission, it shouldn't be too hard even for us developers. 

What are the most minimal steps you envision would be needed ? 

Will a "Mission Wizard" be possible for creating a brand new mission type, in which we only have to "slap in" a few UI elements together and connecting them so that you have both inputs and outputs ?

**Comment by bevans2000** [comment]
Yes I think that is feasible for custom missions to be created. I can see a path to that go now but there are number of steps; we can capture those in the GitHub project.

One question, In Settlements we limit the range for different Mission Types. I want to drop this as it's unneeded limitation. Since we now have different VehicleSpec, some vehicle of the same type have different ranges. Plus the mission score should be penalising long ranges.

**Comment by mokun** [comment]
> we limit the range for different Mission Types. I want to drop this as it's unneeded limitation. 

Sure. The range is only a reference at best.

The real deal is estimating food, fuel, water, oxygen needed, as you know.

Can we try aggregating the result of each mission to get a better picture of how much the mission risk, given the distance, and the amount of consumable provided ?

It will give us a clearer picture on what works and what doesn't.

**Comment by mokun** [comment]
So, if we remove the mission range, then, there's no needed in drawing the range circle when creating a mission, right ?

**Comment by mokun** [comment]
But, we should provide a mission risk score, when player creates a mission.

The risk score will be based on various factors and is NOT the same as the mission score that was needed for the mission to be approved.

**Comment by bevans2000** [comment]
> So, if we remove the mission range, then, there's no needed in drawing the range circle when creating a mission, right ?

No the circles are based on the vehicle range in the new situation

**Comment by bevans2000** [comment]
> But, we should provide a mission risk score, when player creates a mission.
> 
> The risk score will be based on various factors and is NOT the same as the mission score that was needed for the mission to be approved.

Yes that could work. I think we might need a better way go score the missions but we can revisit that later.

**Comment by mokun** [comment]
> the circles are based on the vehicle range in the new situation

Then we still need to fix that. Any ideas ?

Of course the situation is that the resolution is now more fine grain and the range circle gets bigger and go beyond the size of the map box. 

May be having the ability to zoom in and out of the map would make the range circle go from big to small. 

When the circle gets small enough, it would allow the player to have the full range to drag around to stake and pinpoint some locations for explorarion/prospecting.

**Comment by bevans2000** [comment]
This proving to be a big job and shows the limitation of having a fixed set of missions.
I'm going to replace MissionType with a MissionCategory that has less values but these will be applicable to many different types of Missions including custom missions eventually. It will also be used in the Person experience context as well.

**Comment by bevans2000** [comment]
I plan to create a pull request on the changes I have done already sometime this weekend. The remaining areas to visit are

- Trade Deal calculations
- Mission experience in Person
- UI mission create logic

The last one needs some thought because we need to resolve this to support custom missions.

**Comment by bevans2000** [comment]
I've got as far as I can with this change. Totally removing ```MissionType``` is not possible until we fully implement #837. The type is needed to drive the custom UI screens to create the missions.
However this change has removed the need to use ```MissionType``` in the running of a Mission; i.e. the new Mission implementation can define a null type. The MissionType can not be considered the "template" the Mission follows.
The next step is to create a basic driving mission using the new approach, see #890.
